Use available virtual time automatically, harmonize time measurement

OliverO2 opened the pull request:
Provide testTimeSource() which automatically returns virtual time in
(test) coroutines configured to use it, TimeSource.Monotonic otherwise.

Contains stabilization fixes for
* GlobalTimeoutTest
* EngineTimeoutTest
* LongHttpRequestTest
* engine.concurrency.TimeoutTest
* SpecInvocationTimeoutShouldNotApplyToContainersTest
* TestTimeoutOverridesSpecFunctionTest
* EventuallyTest
* EventuallySpec

OliverO2 committed:
Use available virtual time automatically, harmonize time measurement

Provide testTimeSource() which automatically returns virtual time in
(test) coroutines configured to use it, TimeSource.Monotonic otherwise.

OliverO2 committed:
Relax timing in LongHttpRequestTest, engine.concurrency.TimeoutTest

OliverO2 committed:
Revert "Stabilize GlobalTimeoutTest and EngineTimeoutTest (#4124)"

This reverts commit 22d38d89a5afb3fb99fa4db1150b7083ef565562.

OliverO2 committed:
Relax timing in GlobalTimeoutTest

OliverO2 committed:
Relax timing in EngineTimeoutTest

aSemy commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/GlobalTimeoutTest.kt:
```diff
@@ -1,40 +1,39 @@
-      test("global timeouts should apply if no other timeout is set") {
-         val c = ProjectConfiguration().apply { timeout = 3 }
+      test("global timeouts should apply if no other timeout is set")
+      {
```
could you reformat this file please?

OliverO2 replied:
Sure. Just waiting for the re-run to complete before pushing again.

aSemy commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/GlobalTimeoutTest.kt:
```diff
@@ -1,40 +1,39 @@
-         collector.result("suspend").asClue { result -> result?.isError shouldBe true }
+         collector.tests.mapKeys { it.key.name.testName }["blocked"]?.isError shouldBe true
+         collector.tests.mapKeys { it.key.name.testName }["suspend"]?.isError shouldBe true
       }
```
Can the assertions be kept? They're more useful than the ones in this PR.

OliverO2 replied:
You mean these?
```kotlin
         collector.names.shouldContainExactly("blocked", "suspend")

         collector.result("blocked").asClue { result -> result?.isError shouldBe true }
         collector.result("suspend").asClue { result -> result?.isError shouldBe true }
```

No problem.

OliverO2 committed:
Relax timing in SpecInvocationTimeoutShouldNotApplyToContainersTest

OliverO2 committed:
GlobalTimeoutTest: reformat, reestablish exceptions, increase timeout

OliverO2 committed:
Relax timing in TestTimeoutOverridesSpecFunctionTest

aSemy commented on kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/assertions/timing/ContinuallyTest.kt:
```diff
@@ -1,51 +1,63 @@
+
+      "continually" should {
+         "pass working tests" {
+            continually(500.milliseconds) {
+               nonConstantTrue() shouldBe true
+            }
```
Just to double check, is this PR supposed to make the deprecated `continually` use virtual time? I'm asking because test takes 500ms, but I'm not sure about whether it's supposed to use virtual time or not.

OliverO2 replied:
Good catch! In fact, continually is still using `TimeSource.Monotonic`, so the above is wrong. I'll try to change continually to `testTimeSource()` in order to avoid bothering us with flaky tests while continually is still around.

aSemy commented on kotest-common/src/jvmMain/kotlin/io/kotest/common/nonConstantResults.kt:
Could and/or should `nonConstantTrue()` and `nonConstantFalse()` be hidden with `@KotestInternal`?

OliverO2 replied:
I've thought about that, but these seem quite stable and might be useful for Kotest users. WDYT?

aSemy replied:
Can you think of an example usage why a Kotest user might use it? It'd help me understand the context better.

I would prefer avoiding adding top level functions, to avoid cluttering auto-complete suggestions as much as possible.

aSemy replied:
I think they're really useful, and make Kotest's tests easier to understand. But are these functions necessary for this PR? If not, I'd really appreciate it if they were split out into a separate PR.

OliverO2 replied:
Dropping them completely for this PR would be tedious: There was `System.currentTimeMillis()` all over the place for dual use:
* to actually obtain wall clock time, and
* to obtain some value preventing the compiler from optimizing it away.

I wanted a clear separation between these use cases and also use Kotlin TimeMark throughout.

What I could do is `@KotestInternal`ize them for now and then expose them later. To me they seem like the equivalent of JMH's sink holes.

aSemy replied:
> To me they seem like the equivalent of JMH's sink holes.

I agree, which is why I would be careful about including them 😁. Writing performance tests is hard, and should probably be done with the proper tooling (i.e. kotlinx-benchmark or Lincheck).

aSemy commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/EngineTimeoutTest.kt:
```diff
@@ -18,50 +17,41 @@ class EngineTimeoutTest : FunSpec() {
-         collector.result("a").asClue { result ->
-            result?.errorOrNull?.message shouldBe "sleep interrupted"
-         }
+         collector.result("a")!!.errorOrNull!!.message!! shouldBe "sleep interrupted"
```
Could the assertions here be kept? I think that `collector.result("a").asClue { result -> }` will be more useful if the test fails than `collector.result("a")!!`

OliverO2 replied:
Sure.

aSemy commented on kotest-common/src/commonMain/kotlin/io/kotest/common/TestTimeSourceContextElement.kt:
Nice work de-duping the test time source!

OliverO2 replied:
Was more like synchronizing it with the scheduler used. It would be error-prone to continue doing that individually and lead to accidentally using a real-time source with a virtual-time scheduler and vice versa.

OliverO2 committed:
EngineTimeoutTest: reestablish exceptions

OliverO2 committed:
Use testTimeSource() in remaining places, fix tests

OliverO2 committed:
Merge branch 'refs/heads/master' into oo2/test-time-source

OliverO2 committed:
Make nonConstantTrue(), nonConstantFalse() @KotestInternal

OliverO2 committed:
Bind testTimeSource() directly to the scheduler in use

This avoids seeing virtual time in coroutines using non-test
dispatchers, but launched as children of a test dispatcher.

OliverO2 committed:
Merge branch 'master' into oo2/test-time-source

OliverO2 committed:
EventuallyTest, EventuallySpec: remove remaining real-time effects

OliverO2 committed:
EventuallyTest, EventuallySpec: remove unnecessary OptIn declarations

AlexCue987 commented on kotest-common/src/jvmMain/kotlin/io/kotest/common/nonConstantResults.kt:
```diff
@@ -0,0 +1,9 @@
+
+/** returns `true` while preventing the compiler from optimizing it away. */
+@KotestInternal
+fun nonConstantTrue() = System.currentTimeMillis() > 0L
```
the name `nonConstantTrue` does not seem intuitive to me. how about `alwaysTrueNotOptimizedAway`?

OliverO2 replied:
`NotOptimizedAway` is not guaranteed by the code. So while that's the intention, it is also a bit misleading. In any case, this stuff is internal only and I my view probably not worth another iteration.

AlexCue987 commented on kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/assertions/timing/ContinuallyTest.kt:
```diff
@@ -1,51 +1,63 @@
+               continually(12.hours) {
+                  nonConstantFalse() shouldBe true
+               }
+            }
```
this test verifies "fail broken tests", but does it verify "immediately"?

OliverO2 replied:
It sees an immediate failure, but does not verify that it fails immediately. This should probably be addressed in another PR since it is unrelated to this PR's changes and motivation.

AlexCue987 replied:
> addressed in another PR

this

AlexCue987 committed:
Merge branch 'master' into oo2/test-time-source

AlexCue987 reviewed (approved):
After merging the latest commits from master, one test has failed:
```
9 tests completed, 1 failed
> Task :kotest-tests:kotest-tests-spec-parallelism:jvmTest
After Project Error[jvm]

  Test initializationError[jvm] FAILED

  java.lang.IllegalStateException: Parallel execution failure: Execution time was 3.649012400s
```

should we merge this PR anyway, because it's a huge improvement already?

OliverO2 committed:
Merge branch 'refs/heads/master' into oo2/test-time-source

OliverO2 committed:
Make internal config control virtual time in non-deterministic functions

OliverO2 committed:
Update apiDump

aSemy commented on kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/Spec.kt:
```diff
@@ -305,6 +306,13 @@ abstract class Spec : TestConfiguration() {
+    */
+   @KotestInternal
+   var nonDeterministicTestVirtualTimeEnabled = false
+
```
👍

aSemy replied:
```suggestion
   /**
    * Flag controlling whether virtual time is enabled for non-deterministic test functions (`eventually`, ...).
    * This is for Kotest-internal testing only and should not be used outside Kotest.
    */
   @KotestInternal
   var nonDeterministicTestVirtualTimeEnabled: Boolean = false

```

A minor nitpick of mine: I think it's best to explicitly state the type, especially in non-private functions. imo it's easier to scan and to read, and indicates a stricter contract. WDYT?

aSemy commented on kotest-common/src/commonMain/kotlin/io/kotest/common/testTimeSource.kt:
```diff
@@ -0,0 +1,50 @@
+
+@OptIn(KotestInternal::class)
+private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
+   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```
I think it would be better if these functions were also marked as `@KotestInternal`.

- Enabling them requires using the ``, which is also marked as `@KotestInternal`, so it makes sense to me to also mark the utils as internals.
- They're specifically for Kotest internal testing, and so shouldn't be confused with the regular `testTimeSource()` that most users should be using.

```suggestion
@KotestInternal
suspend fun nonDeterministicTestTimeSource(): TimeSource =
   coroutineContext.nonDeterministicTestCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic

@KotestInternal
private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```

aSemy commented on kotest-common/api/kotest-common.api:
```diff
@@ -14,6 +14,21 @@ public abstract interface annotation class io/kotest/common/JVMOnly : java/lang/
+	public static final fun nonConstantTrue ()Z
+}
+
+public final class io/kotest/common/NonDeterministicTestVirtualTimeEnabled : kotlin/coroutines/CoroutineContext$Element, kotlin/coroutines/CoroutineContext$Key {
```
`NonDeterministicTestVirtualTimeEnabled` is marked as `@KotestInternal`, so I'm confused that it's here. Shouldn't it be excluded from the API dump? 🤔

OliverO2 replied:
The Kotlin BCV (`apiDump`/`apiCheck`) does not know about `@KotestInternal`. That could be changed by [configuring the BCV](https://github.com/Kotlin/binary-compatibility-validator?tab=readme-ov-file#optional-parameters), but would be beyond the scope of this PR.

aSemy replied:
Got it, thanks!

aSemy commented on kotest-common/src/commonMain/kotlin/io/kotest/common/testTimeSource.kt:
```diff
@@ -0,0 +1,50 @@
+
+@OptIn(KotestInternal::class)
+private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
+   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```
I have a KDoc suggestion too: add a reference to the flag that enables these functions.

```suggestion
/**
 * Returns the [TimeSource] used for non-deterministic test functions.
 *
 * This is [TimeSource.Monotonic] or – if separately enabled – virtual time, depending on the scheduler in use.
 *
 * @see io.kotest.core.spec.Spec.nonDeterministicTestVirtualTimeEnabled
 */
@OptIn(KotestInternal::class)
suspend fun nonDeterministicTestTimeSource(): TimeSource =
   coroutineContext.nonDeterministicTestCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic

/**
 * Returns the [TimeSource] used for non-deterministic test functions, or `null` if it is not present.
 *
 * @see nonDeterministicTestTimeSource
 * @see io.kotest.core.spec.Spec.nonDeterministicTestVirtualTimeEnabled
 */
@OptIn(KotestInternal::class)
private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```

OliverO2 replied:
(Overlapping suggestions don't work. Addressing this separately.)

OliverO2 replied:
The cross-reference to `io.kotest.core.spec.Spec.nonDeterministicTestVirtualTimeEnabled` won't work (different module). Better leave it out since it's all internal anyway and is better detected by using IDE lookup?

Adding `@see nonDeterministicTestTimeSource` to the private `nonDeterministicTestCoroutineSchedulerOrNull` pointing to its adjacent call site seems to distract more than it would help.

OliverO2 committed:
Update kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/Spec.kt

Co-authored-by: Adam <897017+aSemy@users.noreply.github.com>

OliverO2 committed:
Update kotest-common/src/commonMain/kotlin/io/kotest/common/testTimeSource.kt

Co-authored-by: Adam <897017+aSemy@users.noreply.github.com>
